src/app/app-common/ImguiInstance.cpp

```diff
@@ -0,0 +1,89 @@
+#include "OpenKneeboard/scope_guard.h"
+
+#include <OpenKneeboard/ImguiInstance.h>
+
+#include <vrperfkit/d3d11_helper.h>
+
+#include <imgui_impl_dx11.h>
+
+namespace OpenKneeboard {
+ImguiInstance::ImguiInstance(const DXResources& dxr, KneeboardState* kbs)
+  : mDXR(dxr), mKneeboard(kbs) {
+  m2DContext = mDXR.mD2DBackBufferDeviceContext;
+
+  auto device = dxr.mD3DDevice.get();
+  winrt::com_ptr<ID3D11DeviceContext> ctx;
+  device->GetImmediateContext(ctx.put());
+
+  // init IMGUI
+  mImguiCtx = ImGui::CreateContext();
+  ImGui::SetCurrentContext(mImguiCtx);
+  ImGui_ImplDX11_Init((ID3D11Device*)device, ctx.get());
+
+  // Create texture buffer
+  D3D11_TEXTURE2D_DESC textureDesc {
+    .Width = 768,
+    .Height = 1024,
+    .MipLevels = 1,
+    .ArraySize = 1,
+    .Format = DXGI_FORMAT_B8G8R8A8_UNORM,
+    .SampleDesc = {1, 0},
+    .Usage = D3D11_USAGE_DEFAULT,
+    .BindFlags = D3D11_BIND_SHADER_RESOURCE | D3D11_BIND_RENDER_TARGET,
+  };
+
+  device->CreateTexture2D(&textureDesc, nullptr, mTexture.put());
+
+  mDXR.mD2DDeviceContext->CreateBitmapFromDxgiSurface(
+    mTexture.as<IDXGISurface>().get(), nullptr, mBitmap.put());
+
+  // we're gonna change render targets and generally screw with the state to
+  // render out IMGUI, these helpers should help us restore the state afterwards
+  vrperfkit::D3D11State state {};
+  vrperfkit::StoreD3D11State(ctx.get(), state);
+  scope_guard restoreState(
+    [&]() { vrperfkit::RestoreD3D11State(ctx.get(), state); });
```

Comment:
I've removed vrperfkit from master, as OpenXR Toolkit has a more complete version of this functionality.

I've wrapped this up as `OpenKneeboard::D3D11::SavedState` - just create an instance, and it'll automatically restore when destructed.

Reply:
changed again to use a D3D11 feature I didn't know about (thanks mbucchia :) ) , but no change to the internal API